xde/src/xde.rs

```diff
@@ -1428,11 +1428,33 @@ unsafe extern "C" fn xde_mc_tx(
     // The device must be started before we can transmit.
     let src_dev = &*(arg as *mut XdeDev);
 
-    // TODO I haven't dealt with chains, though I'm pretty sure it's
-    // always just one.
-    assert!((*mp_chain).b_next.is_null());
     __dtrace_probe_tx(mp_chain as uintptr_t);
+    let Ok(mut chain) = PacketChain::new(mp_chain) else {
+        bad_packet_probe(
+            Some(src_dev.port.name_cstr()),
+            Direction::Out,
+            mp_chain as uintptr_t,
+            c"rx'd packet chain from guest was null",
+        );
+        return ptr::null_mut();
+    };
+
+    // TODO: In future we may want to batch packets for further tx
+    // by the mch they're being targeted to. E.g., either build a list
+    // of chains (u1, u2, port0, port1, ...), or hold tx until another
+    // packet breaks the run targeting the same dest.
+    while let Some(pkt) = chain.next() {
+        xde_mc_tx_one(src_dev, pkt);
+    }
 
+    ptr::null_mut()
+}
+
+#[inline]
+unsafe fn xde_mc_tx_one(
+    src_dev: &XdeDev,
+    pkt: Packet<Initialized>,
+) -> *mut mblk_t {
```

Comment:
The comment below seems out of date, since `mp_chain` is owned by the `PacketChain` above

Reply:
Amended, should be slightly clearer now.